[flang] Implement !DIR$ IVDEP directive

JDPailleux opened the pull request:
This directive tells the compiler to ignore vector dependencies in the following loop and it must be placed before a `do loop`.

Sometimes the compiler may not have sufficient information to decide whether a particular loop is vectorizable due to potential dependencies between iterations and the directive is here to tell to the compiler that vectorization is safe with `parallelAccesses` metadata.

This directive is also equivalent to `#pragma clang loop assume(safety)` in C++

kiranchandramohan commented on flang/lib/Lower/Bridge.cpp:
```diff
@@ -2046,6 +2046,37 @@ class FirConverter : public Fortran::lower::AbstractConverter {
   }
 
+  // Add attribute(s) on operations in fir::DoLoopOp if necessary.
+  void attachAttributesToDoLoopOperations(fir::DoLoopOp &doLoop) {
+    if (!doLoop.getOperation())
+      return;
+    if (auto loopAnnotAttr = doLoop.getLoopAnnotationAttr()) {
+      if (loopAnnotAttr.getParallelAccesses().size()) {
+        mlir::LLVM::AccessGroupAttr accessGroupAttr =
+            loopAnnotAttr.getParallelAccesses().front();
+        for (mlir::Block &block : doLoop.getRegion()) {
+          mlir::ArrayAttr attrs =
+              mlir::ArrayAttr::get(builder->getContext(), {accessGroupAttr});
+          for (mlir::Operation &op : block.getOperations()) {
+            if (fir::StoreOp storeOp = mlir::dyn_cast<fir::StoreOp>(op)) {
+              storeOp.setAccessGroupsAttr(attrs);
+            } else if (fir::LoadOp loadOp = mlir::dyn_cast<fir::LoadOp>(op)) {
+              loadOp.setAccessGroupsAttr(attrs);
+            } else if (hlfir::AssignOp assignOp =
+                           mlir::dyn_cast<hlfir::AssignOp>(op)) {
+              // In some loops, the HLFIR AssignOp operation can be translated
+              // into FIR operation(s) containing StoreOp. It is therefore
+              // necessary to forward the AccessGroups attribute.
+              assignOp.getOperation()->setAttr("access_groups", attrs);
+            } else if (fir::CallOp callOp = mlir::dyn_cast<fir::CallOp>(op)) {
+              callOp.setAccessGroupsAttr(attrs);
+            }
+          }
+        }
+      }
+    }
+  }
```
Do we need to go into operations that are nested inside other operations?

```
subroutine sb(A)
  integer, intent(inout) :: A(:)
  integer :: i
  !dir$ ivdep
  do i=1,100
    if (i .gt. 10) then
      A(i) = A(i) + i
    end if 
  end do
end subroutine
```

JDPailleux replied:
Yes, that's right ! So I also have to go through the operations in depth

kiranchandramohan commented on flang/lib/Optimizer/HLFIR/Transforms/ConvertToFIR.cpp:
```diff
@@ -148,13 +148,17 @@ class AssignOpConversion : public mlir::OpRewritePattern<hlfir::AssignOp> {
+      fir::factory::genScalarAssignment(
+          builder, loc, lhsExv, rhsExv, needFinalization,
+          assignOp.isTemporaryLHS(), accessGroups);
     }
```
There are a few other places in this file where `load` and `store` ops are created, should they also be tagged with the access groups?

JDPailleux replied:
I don't think so. The `access group` applies on `loads` and `stores` from assignments in a loop. But I'll make a new pass on it and add tests in that case if needed.

kiranchandramohan reviewed (commented):
A couple of drive-through comments.

tblah commented on flang/include/flang/Optimizer/HLFIR/HLFIROps.td:
```diff
@@ -21,6 +21,7 @@ include "flang/Optimizer/Dialect/FIRAttr.td"
 include "flang/Optimizer/Dialect/FortranVariableInterface.td"
 include "mlir/Dialect/Arith/IR/ArithBase.td"
 include "mlir/Dialect/Arith/IR/ArithOpsInterfaces.td"
+include "mlir/Dialect/LLVMIR/LLVMAttrDefs.td"
```
nit: What is this needed for?

JDPailleux replied:
No, I don't need it ! Sorry

tblah commented on flang/lib/Lower/Bridge.cpp:
```diff
@@ -2046,6 +2046,37 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+              storeOp.setAccessGroupsAttr(attrs);
+            } else if (fir::LoadOp loadOp = mlir::dyn_cast<fir::LoadOp>(op)) {
+              loadOp.setAccessGroupsAttr(attrs);
+            } else if (hlfir::AssignOp assignOp =
+                           mlir::dyn_cast<hlfir::AssignOp>(op)) {
+              // In some loops, the HLFIR AssignOp operation can be translated
+              // into FIR operation(s) containing StoreOp. It is therefore
+              // necessary to forward the AccessGroups attribute.
+              assignOp.getOperation()->setAttr("access_groups", attrs);
```
Tracking everything that could generate a store sounds difficult to maintain in the long term. I wonder if this should instead be added by a pass late in the pipeline after all of these stores have been generated.

JDPailleux replied:
Adding a pass that checks for all `fir.do_loop` with the `parallelAccesses` attribute and do the thing in this code here? It can be done, yes, and it could actually be simpler to manage yes.

tblah reviewed (commented):
Thanks for your work so far on this

ashermancinelli reviewed (commented):
Thanks! I'm not familiar enough with the access group metadata to give better feedback, but I agree with @tblah in wondering if this should happen later, after store ops have already been created from all the other ops that can produce them.

tblah commented:
@JDPailleux do you have any update on this?

JDPailleux commented:
> @JDPailleux do you have any update on this?

Hi, sorry for the late reply (First I was waiting response from https://github.com/llvm/llvm-project/pull/101045 and after I was unavailable). I'll push an updated version soon.

JDPailleux commented:
Hi @kiranktp, thank you for your reply. I was not aware of the existence of a directive in clang for ivdep? I couldn't find this directive, nor the associated metadata. Do you have the related PR or commit ?

kiranchandramohan commented:
> I was not aware of the existence of this directive in clang for ivdep? I couldn't find this directive, nor the associated metadata. Do you have the related PR or commit ?

This is probably a downstream implementation.

JDPailleux commented:
Hi @kiranktp, thanks for your reply.
I've all the information I need to integrate this into Flang, thank you! Let me know when you're done.

JDPailleux committed:
[mlir][LLVMIR] Adding `ivdepEnable` parameter in LoopVectorizeAttr

`loop.vectorize.ivdep.enable` metadata has been added to LLVM to
ignoring vector dependencies. Adding this parameter to LoopVectorizeAttr
will allow Flang to use this metadata with the `!DIR$ IVDEP` directive.

JDPailleux committed:
[flang] Implement !DIR$ IVDEP directive

JDPailleux commented:
Hi, I updated the PR to match what @kiranktp is going to propose.

The first commit is the addition of the `ivdepEnable` parameter in the LoopVectorizeAttr attribute in MLIR. (Maybe to be done in another PR? But only once Kiran has merged what he has done ?).

tblah commented:
Is the ivdep attribute actually supported in upstream llvm? I don't see it in the upstream langref so that makes it hard to judge the correctness of this patch. A quick grep of my llvm source doesn't show anything that looks like support for this.

The changes in this PR look good to me on the flang side. Thank you for contributing @JDPailleux ! I think this patch looks correct to Kiran's description but I wonder if that interface could change whilst Kiran's changes are upstreamed.

@kiranktp are you able to upstream the implementation of this as well as the tests and documentation in [4e0dd91](https://github.com/kiranktp/llvm-project/commit/4e0dd91f213398e36da6837296ece7a8db21b7f2)? If possible I think that needs to land upstream before this patch.

JDPailleux committed:
Revert "[mlir][LLVMIR] Adding `ivdepEnable` parameter in LoopVectorizeAttr"

This reverts commit c4ca0c67f9f720d22a6f6de61a7a58ddb01c5611.

JDPailleux committed:
Updating IVDEP directive to use parallel_access groups + instead of ivdep attr

tblah reviewed (commented):
Just one minor question, otherwise this LGTM. Thanks!

JDPailleux committed:
Udapte attachAccessGroupAttrToLoopOperations with doLoop.walk and uses of all parallelAccesses attr

tblah reviewed (approved):
Thanks!
